OPAL/tac/src/main/scala/org/opalj/tac/fpcf/analyses/cg/AndroidICCAnalysis.scala

```diff
@@ -0,0 +1,1413 @@
+                    if (filterMap(f.componentType)(action).contains(f.categories)) {
+                        filterMap(f.componentType)(action)(f.categories).append(f)
+                    } else {
+                        filterMap(f.componentType)(action) += (f.categories -> ListBuffer(f))
+                    }
+                } else {
+                    filterMap(f.componentType) += (action -> mutable.Map(f.categories -> ListBuffer(f)))
+                }
+            }
+        }
+
+        //match sink and source (intents returned from methods to methods that send the respective intents)
+        senderMap.keySet.foreach { k ⇒
+            if (returningIntentsMap.contains(k)) {
+                val intents = returningIntentsMap(k)
+                implicitIntents --= intents
+                val sender = senderMap(k)
+                intents.foreach { i ⇒
+                    sender.foreach { s ⇒
+                        val intent = i.clone()
+                        intent.caller = s._1
+                        intent.componentTypes += s._2
+                        intent.pc = s._3
+                        implicitIntents += intent
+                    }
+                }
+            } else {
+                //overapproximate (no source found)
+                senderMap(k).foreach{s =>
+                    val intent = new implicitIntent(s._1)
+                    intent.componentTypes += s._2
+                    intent.pc = s._3
+                    intent.action = undef
+                    intent.categories += undef
+                    intent.iData = undef
+                    implicitIntents += intent
+                }
+            }
+        }
+
+        //Intent matching: testing implicitIntents against comptype, actions and categories of intentFilters
+        for (intent ← implicitIntents) {
+            intent.componentTypes.foreach { compType ⇒
+                if (filterMap.contains(compType)) {
+                    if (filterMap(compType).contains(intent.action)) {
+                        filterMap(compType)(intent.action).keySet.foreach { categories ⇒
+                            if ((intent.categories.isEmpty && categories.isEmpty) || intent.categories.forall(categories.contains)) {
+                                if (intent.iData == undef) {
+                                    overApproximateEdges(intent, filterMap(compType)(intent.action)(categories))
+                                } else {
+                                    testData(intent, filterMap(compType)(intent.action)(categories))
+                                }
+                            }
+                        }
+                    }
+                }
+            }
+        }
+        //adding explicit intent edges
+        explicitIntents.foreach { i ⇒
+            val caller = i.caller
+            i.calledMethod.foreach { m ⇒
+                edges += ((caller, i.pc) -> m)
+            }
+        }
+    }
+
+    def testData(
```

Comment:
Please document methods like these. Also, a more meaningful naming of methods seems to be possible. It is totally unclear to me what, e.g., testData, would mean or do.